catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,434 @@
+
+      // Unbind orphaned entries in the map caused by when sources are modified and their key is
+      // different
+      final Map<SourceKey, Source> sourceKeyMap =
+          sources
+              .stream()
+              .collect(
+                  Collectors.toMap(
+                      SourceKey::new,
+                      source -> source,
+                      (source, anotherSourceWithTheSameSourceKey) -> {
+                        LOGGER.warn(
+                            "Found different sources with the same SourceKey with id={}. This means that the SourcePoller might not report accurate availability for these sources in the CatalogFramework. Confirm that every source has a unique id, and try restarting.",
+                            source.getId());
+                        return source;
+                      }));
+
+      final Set<SourceKey> currentSourceKeys = sourceKeyMap.keySet();
+      final Set<SourceKey> sourceKeysInTheCache = cache.asMap().keySet();
+
+      final Set<SourceKey> sourceKeysToRemove = new HashSet<>(sourceKeysInTheCache);
+      sourceKeysToRemove.removeAll(currentSourceKeys);
+      if (CollectionUtils.isNotEmpty(sourceKeysToRemove)) {
+        LOGGER.debug(
+            "Found {} SourceKeys to remove from the cache. This is caused by when sources are destroyed or modified in a way that their SourceKey changes. Removing these from the cache now",
+            sourceKeysToRemove.size());
+        sourceKeysToRemove.forEach(cache::invalidate);
+      }
+
+      Map<SourceKey, Future<SourceAvailability>> availabilityFutures =
+          startAvailabilityChecks(sourceKeyMap);
+      CacheResults(availabilityFutures);
+    } catch (Throwable throwable) {
+      LOGGER.warn("Scheduled polling of sources failed", throwable);
+    }
+  }
+
+  private Map<SourceKey, Future<SourceAvailability>> startAvailabilityChecks(
+      final Map<SourceKey, Source> sourceKeyMap) {
+    Map<SourceKey, Future<SourceAvailability>> availabilityFutures = new HashMap<>();
+    for (final Entry<SourceKey, Source> entry : sourceKeyMap.entrySet()) {
+      final Source source = entry.getValue();
+      final SourceKey sourceKey = entry.getKey();
 
-    this.runner = incomingRunner;
+      boolean sourceIsNotInCache = cache.getIfPresent(sourceKey) == null;
+      if (sourceIsNotInCache) {
+        LOGGER.debug("Found a new SourceKey for source id={}", source.getId());
+        cache.put(sourceKey, SourceAvailability.unknown());
+      }
 
-    scheduler =
-        Executors.newSingleThreadScheduledExecutor(
-            StandardThreadFactoryBuilder.newThreadFactory("sourcePollerThread"));
+      LOGGER.trace("Starting a process to check the availability of source id={}", source.getId());
+      Future<SourceAvailability> sourceAvailabilityFuture =
+          Futures.withTimeout(
+              sourceAvailabilityChecksThreadPool.submit(() -> checkAvailability(source)),
+              DEFAULT_TIMEOUT,
```

Comment:
❓Should we use a timeout that is a few seconds shorter than the polling interval and at most 2 minutes (or something along those lines) instead of hard-coding to 30 seconds? My concern with 30 seconds is that it might be too short for certain sources if the network connectivity is poor, in which case those sources will always be reported as unavailable even though they might actually be available if had waited longer.

Reply:
If we want to use the callback mentioned above I think that would be a good idea. If we don't, however, then we may have to wait for the timeout to get the availability of all our resources so we should probably keep it short.